### PR TITLE
Add MaxRedirects option

### PR DESCRIPTION
**Pull request by intrip**
Add `MaxRedirects` option to set maximum redirection-followings allowed.
The option is only valid when `FollowRedirects` is `true`.

Being able to limit the amount of redirections is helpful in order to avoid possible loops of redirections or just too long round trips.

### CODE REVIEW

**Comment by intrip** [comment]
CI linter issue looks unrelated 🤷‍♂️.

File: cmd/imageproxy/main.go [hunk 1]
```diff
@@ -39,6 +39,7 @@ var denyHosts = flag.String("denyHosts", "", "comma separated list of denied rem
 var referrers = flag.String("referrers", "", "comma separated list of allowed referring hosts")
 var includeReferer = flag.Bool("includeReferer", false, "include referer header in remote requests")
 var followRedirects = flag.Bool("followRedirects", true, "follow redirects")
+var maxRedirects = flag.Uint("maxRedirects", 20, "maximum redirection-followings allowed: 0-254 range. 0 is no limit")
```
Comment (intrip) [inline cmd/imageproxy/main.go:42]: I'm not totally sure what a good default is for this case, I though `20` is ok as it's similar of browser's defaults. We could even go with a `0` to be backward compatible (by default 😂).

File: cmd/imageproxy/main.go [hunk 2]
```diff
@@ -87,6 +88,7 @@ func main() {
 
 	p.IncludeReferer = *includeReferer
 	p.FollowRedirects = *followRedirects
+	p.MaxRedirects = uint8(*maxRedirects)
```
Comment (intrip) [inline cmd/imageproxy/main.go:91]: `254` looks a reasonable max value for me considering that we can also just set `0` to disable the limiting for good.

File: imageproxy.go [hunk 3]
```diff
@@ -185,6 +190,13 @@ func (p *Proxy) serveImage(w http.ResponseWriter, r *http.Request) {
 	if p.FollowRedirects {
 		// FollowRedirects is true (default), ensure that the redirected host is allowed
 		p.Client.CheckRedirect = func(newreq *http.Request, via []*http.Request) error {
+			if p.MaxRedirects > 0 && uint8(len(via)) > p.MaxRedirects {
+				if p.Verbose {
+					p.logf("followed too many redirects: %d", len(via))
+				}
+				http.Error(w, errTooManyRedirects.Error(), http.StatusBadRequest)
```
Comment (intrip) [inline imageproxy.go:197]: I really don't know what good response status is for this scenario so I just went for the catch-all `400` for now 🤷‍♂️

**Comment by intrip** [comment]
@rosa can you take a peek 🙏 ? This should hopefully handle the [carefour issue](https://3.basecamp.com/2914079/buckets/27/todos/4342207110).

File: cmd/imageproxy/main.go (hunk 1 above)
Comment (rosa) [reply cmd/imageproxy/main.go:42]: 20 sounds great to me, although to upstream this I think 0 would make more sense to be backwards compatible 😅 I don't think it's a big deal, you can try with 20 for the upstream PR as well! If they're ok with it, then perfect!

**Review by rosa — APPROVED** [review]
This looks great to me! After shipping the change for us, do open a PR in https://github.com/willnorris/imageproxy, I'm sure they'll like it.

**Pull request merged**